### PR TITLE
Add move/insert to repeated protomodel & add shader and timeline editors

### PR DESCRIPTION
**Pull request by fundies**

### CODE REVIEW

File: Widgets/StackedCodeWidget.cpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+#include "CodeWidget.h"
+
+#include <QFileDialog>
+#include <QInputDialog>
+#include <QMessageBox>
+#include <QTextStream>
+
+StackedCodeWidget::StackedCodeWidget(QWidget* parent) :  QStackedWidget(parent) {}
+
+void StackedCodeWidget::newSource() {
```
Comment (JoshDreamland) [inline Widgets/StackedCodeWidget.cpp]: Does this happen a lot? Consider extracting a template that delegates methods to its type.

File: Editors/PathEditor.h [hunk 2]
```diff
@@ -12,7 +12,7 @@ class PathEditor : public BaseEditor {
 
  public:
   explicit PathEditor(ProtoModelPtr model, QWidget* parent);
-  ~PathEditor();
+  ~PathEditor() override;
```
Comment (JoshDreamland) [inline Editors/PathEditor.h]: This is a very positive style of change... it'd be nice if we could get clang-tidy checks for stuff like this. Are we building with -Wall?

File: Editors/RoomEditor.cpp [hunk 3]
```diff
@@ -70,19 +65,35 @@ RoomEditor::RoomEditor(ProtoModelPtr model, QWidget* parent) : BaseEditor(model,
...
+  // track the cursor position for the status bar
+  ui->roomPreview->widget()->installEventFilter(this);
+  // track the cursor position on the transparency pattern for the status bar
+  ui->roomPreview->widget()->setMouseTracking(true);
+  // also need to track it on the room preview itself so the event is received
+  ui->roomView->setMouseTracking(true);
+
+  cursorPositionLabel = new QLabel();
+  this->updateCursorPositionLabel(ui->roomPreview->widget()->cursor().pos());
+  assetNameLabel = new QLabel("obj_xxx");
```
Comment (JoshDreamland) [inline Editors/RoomEditor.cpp]: Is this a placeholder?

File: Editors/SoundEditor.cpp [hunk 4]
```diff
@@ -114,7 +121,8 @@ void SoundEditor::on_loadButton_clicked() {
         // QString newData = GetModelData(Sound::kDataFieldNumber).toString();
         // TODO: Copy data into our egm and reset the path
         // SetModelData(Sound::kDataFieldNumber, lastData);
-      } else qDebug() << "Failed to load gmx sound";
+      } else
+        qDebug() << "Failed to load gmx sound";
```
Comment (JoshDreamland) [inline Editors/SoundEditor.cpp]: Did clang-format do this? It should have put {} around this else clause.

File: Models/ProtoModel.cpp [hunk 5]
```diff
@@ -101,37 +107,22 @@ bool ProtoModel::setData(const QModelIndex &index, const QVariant &value, int ro
...
-    case CppType::CPPTYPE_DOUBLE:
-      refl->SetDouble(protobuf, field, value.toDouble());
-      break;
-    case CppType::CPPTYPE_FLOAT:
-      refl->SetFloat(protobuf, field, value.toFloat());
-      break;
-    case CppType::CPPTYPE_BOOL:
-      refl->SetBool(protobuf, field, value.toBool());
-      break;
+    case CppType::CPPTYPE_INT32: refl->SetInt32(protobuf, field, value.toInt()); break;
```
Comment (JoshDreamland) [inline Models/ProtoModel.cpp]: I am struggling to believe that clang-format did this, but I don't mind it.

File: Editors/TimelineEditor.cpp [hunk 6]
```diff
@@ -1,10 +1,175 @@
...
+  connect(ui->addMomentButton, &QPushButton::pressed, [=]() {
+    AddMoment(ui->stepBox->value());
+    codeEditor->setDisabled(false);
+    SetCurrentEditor(momentsModel->rowCount() - 1);
+    ui->stepBox->setValue(ui->stepBox->value() + 1);
+  });
+
+  connect(ui->stepBox, QOverload<int>::of(&QSpinBox::valueChanged), [=](int value) { CheckDisableButtons(value); });
+
+  connect(ui->changeMomentButton, &QPushButton::pressed, [=]() {
```
Comment (JoshDreamland) [inline Editors/TimelineEditor.cpp]: Some of these lambdas are big enough to extract into members.

File: Editors/SoundEditor.cpp (hunk 4 above)
Comment (fundies) [reply Editors/SoundEditor.cpp]: it did but i dont think clang format puts brackets

File: Editors/RoomEditor.cpp (hunk 3 above)
Comment (fundies) [reply Editors/RoomEditor.cpp]: yes that was there before. robert wrote that

File: Editors/PathEditor.h (hunk 2 above)
Comment (fundies) [reply Editors/PathEditor.h]: nope.

--   'CMAKE_C_FLAGS_DEBUG': -g
--   'CMAKE_C_FLAGS_MINSIZEREL': -Os -DNDEBUG
--   'CMAKE_C_FLAGS_RELEASE': -O3 -DNDEBUG
--   'CMAKE_C_FLAGS_RELWITHDEBINFO': -O2 -g -DNDEBUG
--   'CMAKE_CXX_FLAGS_DEBUG': -g -fno-omit-frame-pointer -fsanitize=address
--   'CMAKE_CXX_FLAGS_MINSIZEREL': -Os -DNDEBUG
--   'CMAKE_CXX_FLAGS_RELEASE': -O3 -DNDEBUG
--   'CMAKE_CXX_FLAGS_RELWITHDEBINFO': -O2 -g -DNDEBUG

**Pull request merged**